lib/msf/core/auxiliary/scanner.rb

```diff
@@ -108,17 +108,20 @@ def run
         # Stop scanning if we hit a fatal error
         break if has_fatal_errors?
 
-        host = ar.next_host
-        break if not host
+        begin
+          datastore = rhosts_walker.next
+        rescue StopIteration
+          datastore = nil
+        end
+        break unless datastore
 
-        @tl << framework.threads.spawn("ScannerHost(#{self.refname})-#{host[:address]}", false, host.dup) do |thr_host|
-          targ = thr_host[:address]
+        @tl << framework.threads.spawn("ScannerHost(#{self.refname})-#{datastore['RHOST']}", false, datastore.dup) do |thr_datastore|
+          targ = thr_datastore['RHOST']
           nmod = self.replicant
-          nmod.datastore['RHOST'] = thr_host[:address]
-          nmod.datastore['VHOST'] = thr_host[:hostname] if nmod.options.include?('VHOST') && nmod.datastore['VHOST'].blank?
```

Comment:
Unrelated to the PR, but catching up with @smcintyre-r7 it'd be good to update https://github.com/rapid7/metasploit-framework/wiki/How-to-use-datastore-options with vhost/targeturi/rport/etc.